### PR TITLE
sshfs: update so this will build on Linux

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

A user has opened this PR in linuxbrew-core which I'd like to figure out how to merge here:  https://github.com/Homebrew/linuxbrew-core/pull/22889.  I explained here why we need to do this: https://github.com/Homebrew/linuxbrew-core/issues/22813#issuecomment-812958714. 

The short version is that the version of `sshfs` we currently have in this formula won't build on Linux anymore because it doesn't work with version of 3 of `libfuse`, while the latest version of `sshfs` does work (hence why the PR passes CI in linuxbrew-core).  

What is our strategy moving forward for updating packages which depend on FUSE?  I have a long list of other FUSE packages that will probably work on Linux with small fixes that will have the same issue.  Should we just start adding `depends_on :linux`?  Or is there some better interim solution?

### CODE REVIEW

**Comment by carlocab** [comment]
If we need to update this, we may as well make it
```
depends_on :linux
```
No bottles were built here, so this formula will essentially be uninstallable on macOS after the version bump.

Of course, users may still be able to build this from source, but we don't support that. They can do `brew extract sshfs` (or use an external tap) if they want to build this from source.

**Comment by carlocab** [comment]
Alternatively, if you don't mind the extra work later on, you can update this in Linuxbrew-core but leave it at the current version here, and save making this `depends_on :linux` as one of the last things done before the cores are merged.

**Comment by danielnachun** [comment]
I think we're not far out from merging the cores so it won't make a big difference if we make this Linux only now or later, so I think I'll just add `depends_on :linux` now.

File: Formula/sshfs.rb [hunk 1]
```diff
@@ -13,34 +12,21 @@ class Sshfs < Formula
...
-    sha256 "ab2aa697d66457bf8a3f469e89572165b58edb0771aa1e9c2070f54071fad5f6"
-  end
+  depends_on "glib"
+  depends_on "libfuse"
+  depends_on :linux # Because FUSE must be installed as a cask, and cask dependencies are not allowed
 
   def install
-    system "./configure", "--disable-dependency-tracking", "--prefix=#{prefix}"
-    system "make", "install"
+    mkdir("build") do
```
Comment (carlocab) [inline Formula/sshfs.rb:24]:
```suggestion
    mkdir "build" do
```
I don't feel too strongly about this, but this is what all the other formulae (with one exception) do, I think.

File: Formula/sshfs.rb [hunk 2]
```diff
@@ -13,34 +12,21 @@ class Sshfs < Formula
...
+      system "meson", ".."
+      system "meson", "configure", "--prefix", prefix
+      system "ninja"
+      system "ninja", "install"
```
Comment (carlocab) [inline Formula/sshfs.rb:28]:
```suggestion
      system "ninja", "--verbose"
      system "ninja", "install", "--verbose"
```
This keeps the logs complete when we need them.

**Comment by carlocab** [comment]
Looks like you need `net-tools` here in homebrew/core first.

**Comment by danielnachun** [comment]
It looks like `net-tools` will need some testing linuxbrew-core first before it can be upstreamed here, so I'm tagging this with `do-not-merge` until that is done.

**Review by MikeMcQuaid — DISMISSED** [review]

File: Formula/sshfs.rb [hunk 3]
```diff
@@ -13,34 +12,21 @@ class Sshfs < Formula
...
-  # recommended to use e.g. `-o iosize=1048576` (or other, reasonable value)
-  # when launching `sshfs`, for improved performance.
-  # See also: https://github.com/libfuse/sshfs/issues/11
-  patch do
-    url "https://github.com/libfuse/sshfs/commit/667cf34622e2e873db776791df275c7a582d6295.patch?full_index=1"
-    sha256 "ab2aa697d66457bf8a3f469e89572165b58edb0771aa1e9c2070f54071fad5f6"
-  end
+  depends_on "glib"
+  depends_on "libfuse"
+  depends_on :linux # Because FUSE must be installed as a cask, and cask dependencies are not allowed
```
Comment (Bo98) [inline Formula/sshfs.rb:21]: More than that even. sshfs 3.x requires libfuse 3.x but macfuse still uses libfuse 2.x.

**Comment by Bo98** [comment]
We disabled this on the macOS side here: https://github.com/Homebrew/homebrew-core/commit/19292aacd940d8e8e723a232d5877361b60c86fe

I would keep it like that on the macOS side for now since this still receives a significant number of downloads. When the `disable!` is removed, they'll all become `depends_on :linux`.

You can of course still update the formula - we don't need to worry about macOS now that it is disabled.

**Comment by danielnachun** [comment]
I think this is ready for final review, now that `net-tools` has been added and I rebased to incorporate the change to disable on macOS.

**Comment by danielnachun** [comment]
And I spoke too soon.  Now that problem is that `libfuse3` assumes it is being built as root, but it no longer is.  Interestingly, in my prior experience, brewed `libfuse3` seems to be non-functional if installed as a non-root user, because FUSE requires kernel level access.  FUSE based tools can use the system `libfuse`, if available, but fail if they use brewed `libfuse3` as a non-root user.

**Comment by carlocab** [comment]
Seems to me that this just makes it impossible for us to support formulae that need `libfuse` then.

**Comment by danielnachun** [comment]
Actually digging in Anaconda's build script I see there is a `-Duseroot=false` option.  I will explore this and see if it resolves the issue.

**Comment by carlocab** [comment]
`libfuse` should be good to go. #75058 Just rebase.

**Review by SMillerDev — APPROVED** [review]
